frontends/p4/sideEffects.cpp

```diff
@@ -436,19 +446,18 @@ class DismantleExpression : public Transform {
                 leftValue = false;
             else
                 leftValue = true;
-            auto paramtype = typeMap->getType(p, true);
             const IR::Expression* argValue;
             visit(arg);  // May mutate arg!  Recursively simplifies arg.
             auto newarg = result->final;
             CHECK_NULL(newarg);
 
             if (useTemp) {
                 // declare temporary variable
-                auto tmp = refMap->newName("tmp");
+                auto paramtype = typeMap->getType(p, true);
+                if (paramtype->is<IR::Type_Dontcare>())
+                    paramtype = typeMap->getType(arg, true);
```

Comment:
you may want to apply getP4Type to the paramType. This will, for example, convert a Type_Struct to a Type_Name that refers to the struct, which is what you should use for a declaration.

Reply:
createTemporary already calls getP4Type on the type passed in -- just trying to remove redundant repeated calls here.